cpp/include/raft/neighbors/ivf_flat_types.hpp

```diff
@@ -200,10 +219,15 @@ struct index : ann::index {
   ~index()                          = default;
 
   /** Construct an empty index. It needs to be trained and then populated. */
-  index(const handle_t& handle, raft::distance::DistanceType metric, uint32_t n_lists, uint32_t dim)
+  index(const handle_t& handle,
+        raft::distance::DistanceType metric,
+        uint32_t n_lists,
+        bool adaptive_centers,
```

Comment:
Rather than making this an option that persists with the index, why not just make it an optional argument on the `extend()` function? That would enable the user to determine whether it should be done outside of the creation of the index. It's really not an index option, but a design detail based on a specific use-case / usage-pattern.

Reply:
Although I see it sounds more intuitive to keep the parameter in the `extend`, I'd disagree for two reasons:

1. The update-centers routine does not go through the whole data (which would be costly), but only aggregates newly added data. For this to work, the `centers()` must always be up-to-date. Hence, we cannot allow doing an "updating" `extend` after a "non-updating" one. Also from the logical point of view, this would break both alternative invariants (centers are neither constant nor up-to-date). 
2. I hope to preserve the public API the same for all our (ANN) models: `build` gets all parameters through the struct, `extend` never accepts extra parameters (and gets  them from the index). Besides, the `build` function would need to take `update_centers` as well, because it optionally calls `extend`.